test/test_broker.hpp

```diff
@@ -900,6 +1096,47 @@ class test_broker {
                 }
             }
         }
+
+        // Re-activate any saved "shared-subscriptions" that match the newly subscribed topics.
+        // if any match, publish any saved messages right away.
+        for (auto const& e : entries) {
+            MQTT_NS::buffer share;
+            MQTT_NS::buffer topic;
+            std::tie(share, topic) = parse_shared_subscription(std::get<0>(e));
+            MQTT_NS::subscribe_options options = std::get<1>(e);
+
+            // If any of the newly added subscriptions matches a shared subscription
+            // in the saved_shared_subs_ list, then we need to "re-activate" that
+            // shared subscription by removing it from the saved_shared_subs_ list.
+            // If there is one in the saved_shared_subs_ list, then this client is now
+            // the only active client subscribed to the shared subscription, and should
+            // be sent any saved messages that were queued for that subscription.
+            if(!share.empty()) {
+                auto & sharedSubsIdx = saved_shared_subs_.get<tag_combined>();
+                auto const& it = sharedSubsIdx.find(std::make_tuple(share, topic));
+                if(it != sharedSubsIdx.end()) {
+                    // Send the saved messages out on the wire.
+                    for (auto & d : it->messages) {
+                        // But *only* for this connection
+                        // Not every connection in the broker.
+                        ep.publish(
+                            as::buffer(topic),
+                            as::buffer(d.contents),
+                            // TODO: why is this 'retain'?
+                            std::min(options.get_qos(), d.qos_value) | MQTT_NS::retain::yes,
+                            *(d.props),
+                            std::make_tuple(topic, d.contents, *(d.props))
+                            );
+                    }
+
+                    // This initializes the round-robin iterator storage for this shared subscription, if and only if, it is not already.
+                    // Notably, we do not check the return of the emplace call, since if the item already existed, we don't want a new one made.
+                    shared_subs_.emplace(share, topic, subs_.get<tag_topic>().find(std::make_tuple(share, topic)));
```

Comment:
This is a related comment of https://github.com/redboltz/mqtt_cpp/pull/672/files#r502181653
The emplace only works if saved_shared_subs_ exists. There is no change to emplace newly shared subscribed one.

Reply:
subs_.get<tag_topic>().find(std::make_tuple(share....))

Will find the subscription for this share-group and topic from the for loop earlier in the function on line 1068

Reply:
I don't understand yet.

1. Client c1 connected to broker with Session Expiry Interval: infinity.
2. Client c1 subscribed t1 with share1 (QoS0)

In this timing, emplacing some information to shared_subs_ is not needed ???

Reply:
In do_publish(), shared_subs_ is referred.
https://github.com/jonesmz/mqtt_cpp/blob/shared-subscription/test/test_broker.hpp#L1243

If the first shared subscribe dosen't add shared_subs_, then the publish message is not delivered.